### ISSUE TITLE
Keras imported model  throws IllegalStateException at ReshapePreprocessor layer: "Input shape ... and target shape ... do not match"

### ISSUE THREAD

**Issue by davidrizo** [deeplearning4j/deeplearning4j#9523]
Hi,

I've imported a Keras file using:

```
KerasModelBuilder builder = new KerasModel().modelBuilder().modelHdf5Filename(<file name>)
        .enforceTrainingConfig(false);
KerasModel model = builder.buildModel();
INDArray result = model.getComputationGraph().outputSingle(indArray);
```

The shape of indArray is (it comes from a grayscale cropped image) after reshaping:
[1, 64, 236, 1]

The model summary is:
```
==============================================================================================================================================================================
VertexName (VertexType)                      nIn,nOut   TotalParams   ParamsShape                                                                      Vertex Inputs          
==============================================================================================================================================================================
the_input (InputVertex)                      -,-        -             -                                                                                -                      
conv1 (ConvolutionLayer)                     1,16       160           W:{16,1,3,3}, b:{1,16}                                                           [the_input]            
batch_normalization_1 (BatchNormalization)   16,16      64            gamma:{1,16}, beta:{1,16}, mean:{1,16}, var:{1,16}                               [conv1]                
leaky_re_lu_1 (ActivationLayer)              -,-        0             -                                                                                [batch_normalization_1]
max1 (SubsamplingLayer)                      -,-        0             -                                                                                [leaky_re_lu_1]        
conv2 (ConvolutionLayer)                     16,32      4.640         W:{32,16,3,3}, b:{1,32}                                                          [max1]                 
batch_normalization_2 (BatchNormalization)   32,32      128           gamma:{1,32}, beta:{1,32}, mean:{1,32}, var:{1,32}                               [conv2]                
leaky_re_lu_2 (ActivationLayer)              -,-        0             -                                                                                [batch_normalization_2]
max2 (SubsamplingLayer)                      -,-        0             -                                                                                [leaky_re_lu_2]        
conv2d_1 (ConvolutionLayer)                  32,64      18.496        W:{64,32,3,3}, b:{1,64}                                                          [max2]                 
batch_normalization_3 (BatchNormalization)   64,64      256           gamma:{1,64}, beta:{1,64}, mean:{1,64}, var:{1,64}                               [conv2d_1]             
leaky_re_lu_3 (ActivationLayer)              -,-        0             -                                                                                [batch_normalization_3]
max_pooling2d_1 (SubsamplingLayer)           -,-        0             -                                                                                [leaky_re_lu_3]        
permute_1 (PreprocessorVertex)               -,-        -             -                                                                                [max_pooling2d_1]      
reshape (PreprocessorVertex)                 -,-        -             -                                                                                [permute_1]            
bidirectional_1 (BidirectionalLayer)         512,64     295.424       fW:{512,256}, fRW:{64,256}, fb:{1,256}, bW:{512,256}, bRW:{64,256}, bb:{1,256}   [reshape]              
dense2 (DenseLayer)                          128,176    22.704        W:{128,176}, b:{1,176}                                                           [bidirectional_1]      
softmax (ActivationLayer)                    -,-        0             -                                                                                [dense2]               
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
            Total Parameters:  341.872
        Trainable Parameters:  341.872
           Frozen Parameters:  0
==============================================================================================================================================================================
```

I'm getting this exception:
```
Caused by: java.lang.IllegalStateException: Input shape [1, 8, 64, 118] and target shape[1, -1, 512] do not match
	at org.deeplearning4j.nn.modelimport.keras.preprocessors.ReshapePreprocessor.preProcess(ReshapePreprocessor.java:111)
	at org.deeplearning4j.nn.graph.vertex.impl.PreprocessorVertex.doForward(PreprocessorVertex.java:62)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputOfLayersDetached(ComputationGraph.java:2438)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(ComputationGraph.java:1792)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputSingle(ComputationGraph.java:1774)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputSingle(ComputationGraph.java:1760)
	at org.deeplearning4j.nn.graph.ComputationGraph.outputSingle(ComputationGraph.java:1638)
```

---------
The same model, loaded using Python is working with the same input shape, and the following summary:
```

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
the_input (InputLayer)       (None, 64, None, 1)       0         
_________________________________________________________________
conv1 (Conv2D)               (None, 64, None, 16)      160       
_________________________________________________________________
batch_normalization_1 (Batch (None, 64, None, 16)      64        
_________________________________________________________________
leaky_re_lu_1 (LeakyReLU)    (None, 64, None, 16)      0         
_________________________________________________________________
max1 (MaxPooling2D)          (None, 32, None, 16)      0         
_________________________________________________________________
conv2 (Conv2D)               (None, 32, None, 32)      4640      
_________________________________________________________________
batch_normalization_2 (Batch (None, 32, None, 32)      128       
_________________________________________________________________
leaky_re_lu_2 (LeakyReLU)    (None, 32, None, 32)      0         
_________________________________________________________________
max2 (MaxPooling2D)          (None, 16, None, 32)      0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 16, None, 64)      18496     
_________________________________________________________________
batch_normalization_3 (Batch (None, 16, None, 64)      256       
_________________________________________________________________
leaky_re_lu_3 (LeakyReLU)    (None, 16, None, 64)      0         
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 8, None, 64)       0         
_________________________________________________________________
permute_1 (Permute)          (None, None, 8, 64)       0         
_________________________________________________________________
reshape (Reshape)            (None, None, 512)         0         
_________________________________________________________________
bidirectional_1 (Bidirection (None, None, 128)         295424    
_________________________________________________________________
dense2 (Dense)               (None, None, 176)         22704     
_________________________________________________________________
softmax (Activation)         (None, None, 176)         0         
=================================================================
Total params: 341,872
Trainable params: 341,648
Non-trainable params: 224
```

-----
I'm using OsX, with this Maven pom.xml

....
        <dl4j-master.version>1.0.0-M1.1</dl4j-master.version>
        <nd4j.backend>nd4j-native</nd4j.backend>
        <javacpp.version>1.5.6</javacpp.version>
        <tensorflow.version>1.15.5</tensorflow.version>

....
<dependency>
            <groupId>org.tensorflow</groupId>
            <artifactId>tensorflow-core-platform</artifactId>
            <version>0.3.3</version>
        </dependency>

        
        <dependency>
            <groupId>org.nd4j</groupId>
            <artifactId>${nd4j.backend}</artifactId>
            <version>${dl4j-master.version}</version>
        </dependency>
        <dependency>
            <groupId>org.deeplearning4j</groupId>
            <artifactId>deeplearning4j-core</artifactId>
            <version>${dl4j-master.version}</version>
        </dependency>
        <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>tensorflow-platform</artifactId>
            <version>${tensorflow.version}-${javacpp.version}</version>
        </dependency>
        <dependency>
            <groupId>org.deeplearning4j</groupId>
            <artifactId>deeplearning4j-modelimport</artifactId>
            <version>${dl4j-master.version}</version>
        </dependency>

...

-----

Just in case, I'm getting these warning:
Warning: Versions of org.bytedeco:javacpp:1.5.4 and org.bytedeco:openblas:0.3.13-1.5.5 do not match.
Warning: Versions of org.bytedeco:javacpp:1.5.4 and org.bytedeco:opencv:4.5.1-1.5.5 do not match.

----
Thank you in advance,
David

**Comment by agibsonccc** [comment]
@davidrizo could you give me the python script to recreate the model? I can assemble it from your print output, but the layer declaration would help a lot for quick reproduction. Thanks!

**Comment by davidrizo** [comment]
Thank you @agibsonccc  for answering so quickly!!

The [code is here](https://github.com/HISPAMUS/end-to-end-recognition/blob/develop/code/training.py)

The method begins at line 194. The mismatch seems at the reshape layer created at line 233.

**Comment by davidrizo** [comment]
I'm sorry, I've just realized the model is not trained with Keras, so I suppose this is the problem :(

**Comment by agibsonccc** [comment]
@davidrizo could you clarify? We support tf keras as well.

**Comment by davidrizo** [comment]
@agibsonccc It is trained with pure TensorFlow in Python, not Keras, and I was using DL4J KerasModel().modelBuilder()

**Comment by agibsonccc** [comment]
@davidrizo you'll want to use our tensorflow import then. Keras has 2 file formats: h5 and pb. Our keras import targets the h5 format. Pb is a normal tensorflow graph which we have a separate API for.

**Comment by davidrizo** [comment]
@agibsonccc Thank you for your advices.

I'm using models trained by other people in my team and there was a misunderstanding on my part with the source code. 

The model used is actually a Keras one (version 2.2.4). The code two build the network is this one:

```python
import numpy as np
import cv2
import os
import itertools
import tensorflow as tf
from keras import backend as K
from keras.layers import Conv2D, MaxPooling2D
from keras.layers import Input, Dense, Activation, BatchNormalization, LeakyReLU
from keras.layers import Reshape, Lambda, Permute
from keras.models import Model
from keras.layers import GRU, LSTM
from keras.layers import TimeDistributed, Bidirectional
from keras.models import load_model


# ===================================================


def ctc_lambda_func(args):
    y_pred, labels, input_length, label_length = args
    y_pred = y_pred[:, :, :]
    return K.ctc_batch_cost(labels, y_pred, input_length, label_length)


def get_model(input_shape, vocabulary_size):

    conv_filters = [16,32,64]

    input_data = Input(name='the_input', shape=input_shape, dtype='float32')

    inner = Conv2D(conv_filters[0], 3, padding='same', name='conv1')(input_data)
    inner = BatchNormalization()(inner)
    inner = LeakyReLU(alpha=0.2)(inner)
    inner = MaxPooling2D(pool_size=(2, 2), name='max1')(inner)

    inner = Conv2D(conv_filters[1], 3, padding='same', name='conv2')(inner)
    inner = BatchNormalization()(inner)
    inner = LeakyReLU(alpha=0.2)(inner)
    inner = MaxPooling2D(pool_size=(2, 1), name='max2')(inner)

    inner = Conv2D(conv_filters[2], 3, padding='same')(inner)
    inner = BatchNormalization()(inner)
    inner = LeakyReLU(alpha=0.2)(inner)
    inner = MaxPooling2D(pool_size=(2, 1))(inner)

    inner = Permute((2, 1, 3))(inner)
    inner = Reshape(target_shape=(-1, (input_shape[0] // (2 ** 3)) * conv_filters[-1]), name='reshape')(inner)

    inner = Bidirectional(LSTM(64, return_sequences = True, dropout=0.25))(inner)

    inner = Dense(vocabulary_size+1, name='dense2')(inner)
    y_pred = Activation('softmax', name='softmax')(inner)

    model_pr = Model(inputs=input_data, outputs=y_pred)
    model_pr.summary()

    labels = Input(name='the_labels',shape=[None], dtype='float32')
    input_length = Input(name='input_length', shape=[1], dtype='int64')
    label_length = Input(name='label_length', shape=[1], dtype='int64')

    loss_out = Lambda(
        ctc_lambda_func, output_shape=(1,),
        name='ctc')([y_pred, labels, input_length, label_length])

    model_tr = Model(inputs=[input_data, labels, input_length, label_length],
                  outputs=loss_out)

    model_tr.compile(loss={'ctc': lambda y_true, y_pred: y_pred}, optimizer='adam')

    # Note: the trained model includes CTC (model_ctc), but we store the one without CTC (model_pr)    
    
    return model_tr, model_pr
```

PS: I've successfully used DL4J to import models in Keras in two other cases.

**Comment by agibsonccc** [comment]
@davidrizo lambda layers won't work without some custom work. Here's a fairly decent example of what you need to do to implement a custom layer. As this is not a bug, I am going to go ahead and re close this. If you want help on your specific model, please do post over on the forums with this as a starting point. https://github.com/eclipse/deeplearning4j/blob/5e8951cd8ee8106bb393635f840c398a1759b2fa/deeplearning4j/deeplearning4j-modelimport/src/test/java/org/deeplearning4j/nn/modelimport/keras/configurations/DeepCTRLambdaTest.java#L109

Note this will use our samediff library for the underlying wrapper there. We do have ctc loss in samediff. You can find it here: 
https://github.com/eclipse/deeplearning4j/blob/fa2059859689393fa3530945d3b9d17ee8ae6343/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/ops/SDLoss.java#L219

Let me know if you have any problems.

**Comment by davidrizo** [comment]
@agibsonccc Thank you !!

**Comment by davidrizo** [comment]
@agibsonccc Having a look to the code more closely, the lambda is not used in the stored model -- see the comment  in the code "Note: the trained model includes CTC (model_ctc), but we store the one without CTC (model_pr)"


Then, the model stored in the h5 used to predict is this one:
```python
conv_filters = [16,32,64]

    input_data = Input(name='the_input', shape=input_shape, dtype='float32')

    inner = Conv2D(conv_filters[0], 3, padding='same', name='conv1')(input_data)
    inner = BatchNormalization()(inner)
    inner = LeakyReLU(alpha=0.2)(inner)
    inner = MaxPooling2D(pool_size=(2, 2), name='max1')(inner)

    inner = Conv2D(conv_filters[1], 3, padding='same', name='conv2')(inner)
    inner = BatchNormalization()(inner)
    inner = LeakyReLU(alpha=0.2)(inner)
    inner = MaxPooling2D(pool_size=(2, 1), name='max2')(inner)

    inner = Conv2D(conv_filters[2], 3, padding='same')(inner)
    inner = BatchNormalization()(inner)
    inner = LeakyReLU(alpha=0.2)(inner)
    inner = MaxPooling2D(pool_size=(2, 1))(inner)

    inner = Permute((2, 1, 3))(inner)
    inner = Reshape(target_shape=(-1, (input_shape[0] // (2 ** 3)) * conv_filters[-1]), name='reshape')(inner)

    inner = Bidirectional(LSTM(64, return_sequences = True, dropout=0.25))(inner)

    inner = Dense(vocabulary_size+1, name='dense2')(inner)
    y_pred = Activation('softmax', name='softmax')(inner)

    model_pr = Model(inputs=input_data, outputs=y_pred)
```
